demos/demo_2d_north_sea.py

```diff
@@ -0,0 +1,234 @@
+
+# To extract free surface elevation timeseries at the tide gauges, we add in
+# some :class:`TimeSeriesCallback2D` instances. We need to provide the solver
+# object, the field names to be evaluated, the UTM coordinates and finally the
+# name of each tide gauge.
+
+trans = pyproj.Transformer.from_crs(coordsys.LL_WGS84.srs, UTM_ZONE30.srs)
+for name, (sta_lat, sta_lon) in stations.items():
+    sta_x, sta_y = trans.transform(sta_lon, sta_lat)
+    cb = TimeSeriesCallback2D(
+        solver_obj,
+        ["elev_2d"],
+        sta_x,
+        sta_y,
+        name,
+        append_to_log=False,
+    )
+    solver_obj.add_callback(cb)
+
+# We still need to add a crucially important component to our tidal model...
+# the tides! To do this, we make use of the
+# `TPXO tidal forcing data set <https://www.tpxo.net/>`__.
+# In order for this demo to work you will need to obtain NetCDF files for
+# the forcing data as described on the
+# `TPXO access page <https://www.tpxo.net/tpxo-products-and-registration>`__.
+# We recommend that you store them in a subdirectory `tpxo` of a directory
+# either located at `~/data` or referenced by the environment variable `$DATA`.
+#
+# With the data in place, we can set up a Firedrake :class:`Function` to control
+# the elevation forcings on ocean boundaries and pass them into a
+# :class:`TPXOTidalBoundaryForcing` instance.
+
+data_dir = os.path.join(os.environ.get("DATA", "~/data"), "tpxo")
+if not os.path.exists(data_dir):
+    raise IOError(f"Forcing data directory {data_dir} does not exist")
+forcing_constituents = ["Q1", "O1", "P1", "K1", "N2", "M2", "S2", "K2"]
+elev_tide_2d = Function(P1_2d, name="Tidal elevation")
+P1v_2d = solver_obj.function_spaces.P1v_2d
+uv_tide_2d = Function(P1v_2d, name="Tidal transport")
+tbnd = forcing.TPXOTidalBoundaryForcing(
+    elev_tide_2d,
+    start_date,
+    UTM_ZONE30,
+    uv_field=uv_tide_2d,
+    data_dir=data_dir,
+    constituents=forcing_constituents,
+    boundary_ids=[100],
+)
+
+# Set the tidal field at time zero (of the simulation).
+tbnd.set_tidal_field(0.0)
+
+# As mentioned above, the forcing data drives the boundary conditions on
+# boundary segments with tag 100. We pass this information to the solver
+# object as follows:
+
+uv_2d, elev_2d = split(solver_obj.fields.solution_2d)
+depth_2d = solver_obj.depth.get_total_depth(elev_2d)
+solver_obj.bnd_functions["shallow_water"] = {
+    100: {"elev": elev_tide_2d, "uv": uv_tide_2d / depth_2d},
+}
+
+# Note that we have assumed a fully "spun-up" tidal model here. It is standard
+# practice to "spin-up" the model from a state of rest, slowly introducing the
+# tidal forcings over one or two simulated weeks. For such preparatory runs, we
+# need to modify the boundary condition expressions slightly. See the example in
+# the source code for details on this. After a two week spin-up period, we obtain
+# the following free surface elevation field (as well as a velocity field).
+#
+# .. figure:: north_sea_init.png
```

Comment:
I'd prefer a diverging symmetric colormap for the tides :)

Reply:
Is there one that you would recommend?

Reply:
`"RdBu_r"` is usually ok.